README.md
# pick-me
Generate functions that pick things from lists with specified frequencies.

## Synopsis

```ts
// Let's make a generator of quasi-Polynesian lorem ipsum!

import { pickMe, rando } from 'pick-me-too'

// we want our text reproducible, so we use a seeded random number generator
const rng = rando(1)

// now we specify the frequency of various patterns in our invented language

const vowels = pickMe(
  [
    ['a', 2],
    ['e', 1],
    ['i', 1.5],
    ['o', 1],
    ['u', 1.5],
    ['ā', 0.75],
    ['ē', 0.5],
    ['ī', 0.75],
    ['ō', 0.5],
    ['ū', 0.75],
  ],
  rng,
)
const consonants = pickMe(
  [
    ["'", 5],
    ['h', 1],
    ['p', 3],
    ['t', 3],
    ['k', 3],
    ['b', 1],
    ['d', 1],
    ['g', 1],
    ['n', 3],
    ['r', 1],
  ],
  rng,
)
// our language has simple CV syllables
const syllable = () => `${consonants()}${vowels()}`
const morphemeLength = pickMe(
  [
    [1, 1],
    [2, 3],
    [3, 1],
  ],
  rng,
)
const morpheme = () => {
  let m = ''
  for (let i = 0, lim = morphemeLength(); i < lim; i++) m += syllable()
  return m
}
const reduplication = () => {
  const r = morpheme()
  return `${r}${r}`
}
// we can pick functions, too!
const segment = pickMe(
  [
    [morpheme, 5],
    [reduplication, 1],
  ],
  rng,
)
const wordLength = pickMe(
  [
    [1, 5],
    [2, 3],
    [3, 2],
    [4, 1],
    [5, 1],
  ],
  rng,
)
const word = () => {
  let w = ''
  for (let i = 0, lim = wordLength(); i < lim; i++) w = w + segment()()
  return w
}
const clauseLength = pickMe(
  [
    [1, 1],
    [2, 3],
    [3, 4],
    [4, 4],
    [5, 5],
    [6, 6],
  ],
  rng,
)
const clause = () => {
  let s = []
  for (let i = 0, lim = clauseLength(); i < lim; i++) s.push(word())
  return s.join(' ')
}

const saying = clause()
// => "'ātika kādepāgū bīpīpe pōhopōhopunā'ī bi'a'anū'ā'idi'idi'ōkē'ōkēpetē"
```

## API

### `function pickMe<T>(frequencies: [T, number][] | T[], rando?: () => number): () => T`

Generates a zero-argument function that returns an item from the frequency list
parameter with a probability concordant with its specified frequency.

The required parameter is a list of pairs: an item to pick and its frequency. The "frequency"
is simply a positive number which determines the probability of picking one item relative to another.
If the parameters are `[[1, 1], [2, 2]]`, you will get twice as many 2s as 1s. You will get the same
results with `[[1, 0.1], [2, 0.2]]`, `[[1, 10], [2, 20]]`, and `[[1, 0.13], [2, 0.26]]`.

The second, optional parameter is a random number sequence generator. Provide this if you want a reproducible
sequence of selections from your frequency list. If this doesn't concern you, leave it off and take
whatever `Math.random()` gives you.

If the frequency list is not a list of pairs but just a list of items, this is equivalent to a list of
pairs where all the frequencies are the same. (Note, this will not work if the items are all pairs whose
second value is a number. If you want to pick among such pairs you'll need to wrap them in new pairs.)

An item with
a non-positive frequency will be culled from the list. `pickMe` will throw an error if the list is empty.
Likewise, it will throw an error if any item appears more than once in the list. `[]` is
not good. `[[1, 1], [1, 2]]` is no better.

Note, the item returned by `pickMe` is the item you provided, not a copy. If you are picking from a list of
objects and you modify one of them, it will remain modified.

Finally, `pickMe` compiles an efficient picker from the parameters provided. As with other compiled things, such
as regular expressions, you want to do this one time in an initialization phase and then reuse it in loops.
The compilation is not terribly expensive, but skipping it altogether is cheaper.

@example
```ts
import { rando, pickMe } from 'pick-me-too'

// make a random number generator with a seed of 1
const rng = rando(1)
const fooBarBaz = pickMe([["foo", 1], ["bar", 2], ["baz", 3]], rng)
const ar: string[] = []
for (let i = 0; i < 10; i++) ar.push(fooBarBaz())
// => [ 'baz', 'foo', 'baz', 'baz', 'baz', 'bar', 'baz', 'baz', 'bar', 'baz' ]
```

### `function pickMeToo<T>(frequencies: [T, number][] | T[]): (rng: () => number) => () => T`

Generates a function from random number generators to generators of `T`.

This is like `pickMe` -- in fact, `pickMe` delegates to this -- but it factors out the random
number generator. It is useful in a situation where you want to incur the compilation of the
picker only once but you want to plug in different random number sequences on different
occasions.

@example
```ts
import { pickMeToo } from 'pick-me-too'

const fooBarBaz = pickMeToo([["foo", 1], ["bar", 1000], ["baz", 1]])

// minimum probablility generator just returns "foo"
fooBarBaz(() => 0)()
// => "foo"

// maximum probability generator just returns "baz"
fooBarBaz(() => 1)()
// => "baz"
```

### `function rando(seed: number): () => number`

Generates a random number sequence generator.

This is a demonstration generator useful for testing and documentation. Any zero-argument function that returns
numbers will suffice for `pickMe`. If it returns 0 or less, the first item
given to `pickMe` will be picked. If it is 1 or more, the last item will be.

The implementation of `rando` is taken from [this Stack Overflow comment](https://stackoverflow.com/a/47593316/15060051).
There are many others if you don't find this suitable.

@example
```ts
import { rando } from 'pick-me-too'

// make a random number generator with a seed of 1
const rng = rando(1)
const ar: number[] = []
for (let i = 0; i < 3; i++) ar.push(rng())
// => [ 0.6270739405881613, 0.002735721180215478, 0.5274470399599522 ]

// reseed the generator
(rng as any)(Math.random())
// => ??? the random number sequence is still deterministic, but at a completely different place
```

## Acknowledgements

It is possible this package is a duplicate. I didn't try very hard to find out. I just felt like writing
this code and, having written it, I felt like publishing it. It may behoove you to check npmjs carefully
and compare packages.

That being said, `pickMe` should be near optimally fast (if the frequences are strongly skewed, it is
possible to compile a picker that uses fewer numerical comparisons in the typical case). Its unit and
documentation tests are thorough.


CHANGELOG.md
# Changelog

## v2.0.0
- improved efficiency of compiled picker
- ***BREAKING CHANGE***: with this improvement the order of picked items is no longer preserved, so a pseudo random number generator that return 0 or less or 1 or more is no longer necessarily returns the first or last item in the list, respectively
## v1.3.0 *2022-7-23*
- made `rando` reseedable
## v1.2.1 *2022-4-30*
- fixed typo in README
## v1.2.0 *2022-4-17*
- some efficiency improvements in edge cases
## v1.1.1 *2022-4-16*
- fixed some project configuration
## v1.1.0 *2022-4-16*
- factored `pickMeToo` out of `pickMe`

src/__tests__/documentation.test.ts
/** tests for all the code shown in documentation */

import { rando, pickMe, pickMeToo } from '..'

describe('README', () => {
  describe('Synopsis', () => {
    const rng = rando(1)
    const vowels = pickMe(
      [
        ['a', 2],
        ['e', 1],
        ['i', 1.5],
        ['o', 1],
        ['u', 1.5],
        ['ā', 0.75],
        ['ē', 0.5],
        ['ī', 0.75],
        ['ō', 0.5],
        ['ū', 0.75],
      ],
      rng,
    )
    const consonants = pickMe(
      [
        ["'", 5],
        ['h', 1],
        ['p', 3],
        ['t', 3],
        ['k', 3],
        ['b', 1],
        ['d', 1],
        ['g', 1],
        ['n', 3],
        ['r', 1],
      ],
      rng,
    )
    const syllable = () => `${consonants()}${vowels()}`
    const morphemeLength = pickMe(
      [
        [1, 1],
        [2, 3],
        [3, 1],
      ],
      rng,
    )
    const morpheme = () => {
      let m = ''
      for (let i = 0, lim = morphemeLength(); i < lim; i++) m += syllable()
      return m
    }
    const reduplication = () => {
      const r = morpheme()
      return `${r}${r}`
    }
    const segment = pickMe(
      [
        [morpheme, 5],
        [reduplication, 1],
      ],
      rng,
    )
    const wordLength = pickMe(
      [
        [1, 5],
        [2, 3],
        [3, 2],
        [4, 1],
        [5, 1],
      ],
      rng,
    )
    const word = () => {
      let w = ''
      for (let i = 0, lim = wordLength(); i < lim; i++) w = w + segment()()
      return w
    }
    const clauseLength = pickMe(
      [
        [1, 1],
        [2, 3],
        [3, 4],
        [4, 4],
        [5, 5],
        [6, 6],
      ],
      rng,
    )
    const clause = () => {
      let s: string[] = []
      for (let i = 0, lim = clauseLength(); i < lim; i++) s.push(word())
      return s.join(' ')
    }
    const saying = clause()
    test('polynesianish', () => expect(saying).toEqual("'ātika kādepāgū bīpīpe pōhopōhopunā'ī bi'a'anū'ā'idi'idi'ōkē'ōkēpetē"))
  })
})

describe('index.ts', () => {
  describe('pickMe', () => {
    const rng = rando(1)
    const fooBarBaz = pickMe(
      [
        ['foo', 1],
        ['bar', 2],
        ['baz', 3],
      ],
      rng,
    )
    const ar: string[] = []
    for (let i = 0; i < 10; i++) ar.push(fooBarBaz())
    test('pickMe', () => expect(ar).toEqual(['bar', 'baz', 'bar', 'foo', 'foo', 'baz', 'bar', 'bar', 'baz', 'foo']))
  })
  describe('pickMeToo', () => {
    const fooBarBaz = pickMeToo([
      ['foo', 1],
      ['bar', 1000],
      ['baz', 1],
    ])
    test('low probability', () => expect(fooBarBaz(() => 0)()).toEqual('bar'))
    test('high probability', () => expect(fooBarBaz(() => 1)()).toEqual('baz'))
  })
  describe('rando', () => {
    const rng = rando(1)
    const ar: number[] = []
    for (let i = 0; i < 3; i++) ar.push(rng())
    test('rando', () => {
      expect(ar).toEqual([0.6270739405881613, 0.002735721180215478, 0.5274470399599522])
    })
    test('rando(1)()', () => expect(rando(1)()).toEqual(0.6270739405881613))
    test('(rando(1) as any)(2)', () => expect((rando(1) as any)(2)).toEqual(0.7342509443406016))
  })
})


src/__tests__/functionality.test.ts
import { pickMe, rando } from '..'

// we need a fixed random number sequence for testing
const rng = rando(1)

describe('rando', () => {
  const bins: number[] = []
  for (let i = 0; i < 100; i++) {
    bins[i] = 0
  }
  for (let i = 0, lim = bins.length * 1000; i < lim; i++) {
    let idx = Math.floor(rng() * 100)
    if (idx === 100) idx = 99
    bins[idx] += 1
  }
  for (let i = 0; i < bins.length; i++) {
    test('even distribution', () => expect(Math.round(bins[i] / 1000)).toEqual(1))
  }
  for (let i = 0; i < 10; i++) {
    const seed = Math.random()
    const rng1 = rando(seed),
      rng2 = rando(seed)
    const ar1: number[] = [],
      ar2: number[] = []
    for (let j = 0; j < 10; j++) {
      ar1.push(rng1())
      ar2.push(rng2())
    }
    test(`two randos with seed ${seed} produce the same sequence of random values`, () => expect(ar1).toEqual(ar2))
    test(`randos can be reseeded`, () => expect(rng1()).not.toEqual((rng2 as any)(Math.random())))
  }
})

describe('pickMe', () => {
  const trueFalse = pickMe(
    [
      ['true', 1],
      ['false', 1],
    ],
    rng,
  )
  for (let i = 0; i < 10; i++) {
    test('returns either "true" or "false"', () => expect(['true', 'false'].indexOf(trueFalse())).toBeGreaterThan(-1))
  }
  test('requires some items', () =>
    expect(() => {
      pickMe([])
    }).toThrowError())
  test('we allow no duplicates', () =>
    expect(() => {
      pickMe([
        ['foo', 1],
        ['foo', 1],
      ])
    }).toThrowError())
  let f = pickMe([['foo', 1]], rng)
  let ar: string[] = []
  for (let i = 0; i < 100; i++) ar.push(f())
  test("we always get the same thing if there's just one item", () =>
    expect(ar.every((s) => s === 'foo')).toEqual(true))
  f = pickMe([
    ['foo', 1],
    ['bar', 0],
    ['baz', -1],
  ])
  ar.length = 0
  for (let i = 0; i < 100; i++) ar.push(f())
  test('we always get the same thing if only one item has a positive frequency', () =>
    expect(ar.every((s) => s === 'foo')).toEqual(true))
  f = pickMe(
    [
      ['foo', 2],
      ['bar', 1],
    ],
    rng,
  )
  const counts1: Record<string, number> = { foo: 0, bar: 0 }
  for (let i = 0; i < 1000; i++) counts1[f()] += 1
  test('we expect some foos', () => expect(counts1.foo).toBeGreaterThan(0))
  test('we expect some bars', () => expect(counts1.bar).toBeGreaterThan(0))
  test('we get roughly as many of each thing as we expect', () =>
    expect(Math.round(counts1.foo / counts1.bar)).toEqual(2))
  // do this with a bit more rigor
  const letters = 'abcdefghijklmnopqrstuvwxyz'.split('')
  const frequencies: [string, number][] = []
  for (let i = 0; i < letters.length; i++) {
    frequencies.push([letters[i], i + 1])
  }
  const segment = (letters.length * (letters.length + 1)) / 2
  f = pickMe(frequencies, rng)
  let counts2: Record<string, number> = {}
  for (const l of letters) counts2[l] = 0
  for (let i = 0, lim = segment * 1000; i < lim; i++) counts2[f()] += 1
  for (const [l, f] of frequencies) {
    const c = Number(counts2[l])
    test(`we got something for ${l}`, () => expect(c).toBeGreaterThan(0))
    const expectation = f * 1000
    test(`we got roughly as many as we expected for ${l}, with frequency ${f}`, () =>
      expect(Math.round(c / expectation)).toEqual(1))
  }
  const f1 = pickMe(
    [
      ['foo', 1],
      ['bar', 1],
      ['baz', 1],
    ],
    rando(1),
  )
  const f2 = pickMe(['foo', 'bar', 'baz'], rando(1))
  const ar1: string[] = []
  const ar2: string[] = []
  for (let i = 0; i < 100; i++) {
    ar1.push(f1())
    ar2.push(f2())
  }
  test('frequencies are optional if all are the same', () => {
    expect(ar1).toEqual(ar2)
  })
})


lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.rando = exports.pickMeToo = exports.pickMe = void 0;
/**
 * Generates a zero-argument function that returns an item from the frequency list
 * parameter with a probability concordant with its specified frequency.
 *
 * The required parameter is a list of pairs: an item to pick and its frequency. The "frequency"
 * is simply a positive number which determines the probability of picking one item relative to another.
 * If the parameters are `[[1, 1], [2, 2]]`, you will get twice as many 2s as 1s. You will get the same
 * results with `[[1, 0.1], [2, 0.2]]`, `[[1, 10], [2, 20]]`, and `[[1, 0.13], [2, 0.26]]`.
 *
 * The second, optional parameter is a random number sequence generator. Provide this if you want a reproducible
 * sequence of selections from your frequency list. If this doesn't concern you, leave it off and take
 * whatever `Math.random()` gives you.
 *
 * If the frequency list is not a list of pairs but just a list of items, this is equivalent to a list of
 * pairs where all the frequencies are the same. (Note, this will not work if the items are all pairs whose
 * second value is a number. If you want to pick among such pairs you'll need to wrap them in new pairs.)
 *
 * An item with
 * a non-positive frequency will be culled from the list. `pickMe` will throw an error if the list is empty.
 * Likewise, it will throw an error if any item appears more than once in the list. `[]` is
 * not good. `[[1, 1], [1, 2]]` is no better.
 *
 * Note, the item returned by `pickMe` is the item you provided, not a copy. If you are picking from a list of
 * objects and you modify one of them, it will remain modified.
 *
 * Finally, `pickMe` compiles an efficient picker from the parameters provided. As with other compiled things, such
 * as regular expressions, you want to do this one time in an initialization phase and then reuse it in loops.
 * The compilation is not terribly expensive, but skipping it altogether is cheaper.
 *
 * @example
 * ```ts
 * import { rando, pickMe } from 'pick-me'
 *
 * // make a random number generator with a seed of 1
 * const rng = rando(1)
 * const fooBarBaz = pickMe([["foo", 1], ["bar", 2], ["baz", 3]], rng)
 * const ar: string[] = []
 * for (let i = 0; i < 10; i++) ar.push(fooBarBaz())
 * // => [ 'baz', 'foo', 'baz', 'baz', 'baz', 'bar', 'baz', 'baz', 'bar', 'baz' ]
 * ```
 *
 * @param {[T, number][] | T[]} frequencies - phrases to match
 * @param {Rng} [rando=() => Math.random()] - random number generator
 * @returns {() => T} a picker of T's
 */
function pickMe(frequencies, rando) {
    const rng = rando !== null && rando !== void 0 ? rando : (() => Math.random());
    return pickMeToo(frequencies)(rng);
}
exports.pickMe = pickMe;
/**
 * Generates pickers of things. This separates the compilation step from frequencies so you
 * can plug different random number generators in.
 *
 * @example
 * ```ts
 * import { pickMeToo } from 'pick-me'
 *
 * const fooBarBaz = pickMeToo([["foo", 1], ["bar", 1000], ["baz", 1]])
 *
 * // minimum probablility generator just returns "foo"
 * fooBarBaz(() => 0)()
 * // => "foo"
 *
 * // maximum probability generator just returns "baz"
 * fooBarBaz(() => 1)()
 * // => "baz"
 * ```
 *
 * @template T - the sort of thing to pick
 * @param {[T, number][] | T[]} frequencies - things to pick according to their frequencies
 * @returns {(rng: Rng) => () => T} - a function from random number generators to pickers of things in the frequency list
 */
function pickMeToo(frequencies) {
    const simpleton = simplePicker(frequencies);
    if (simpleton)
        return simpleton;
    frequencies = sanitize(frequencies);
    // short-circuit edge cases
    if (frequencies.length === 0)
        throw new Error(`nothing to select in ${frequencies}`);
    if (frequencies.length === 1) {
        const [t] = frequencies[0];
        return (_rng) => () => t;
    }
    dupChecker(frequencies, (item) => item[0]);
    let total = 0, acc = 0;
    for (const [, n] of frequencies)
        total += n;
    for (let i = 0; i < frequencies.length; i++) {
        const [, n] = frequencies[i];
        acc += n;
        frequencies[i][1] = acc / total;
    }
    const items = [], thresholds = [];
    for (const [t, n] of frequencies) {
        items.push(t);
        thresholds.push(n);
    }
    const f = thresholder(thresholds);
    return (rng) => () => items[f(rng())];
}
exports.pickMeToo = pickMeToo;
// throw an error in case of duplicates
// it is difficult to do this efficiently because we allow T to be anything, so we can't for instance use a Set
function dupChecker(frequencies, toT) {
    for (let i = 0, lim = frequencies.length - 1; i < lim; i++) {
        const t = toT(frequencies[i]);
        for (let j = i + 1; j < frequencies.length; j++) {
            if (t === toT(frequencies[j]))
                throw new Error(`${t} appears more than once among frequencies`);
        }
    }
}
// if all frequencies are the same, we can make a simpler picker
function simplePicker(frequencies) {
    if (frequencies.length === 0)
        return;
    if (frequencies.every((s) => Array.isArray(s) && s.length === 2 && typeof s[1] === 'number')) {
        const castFrequencies = frequencies;
        const [, n1] = castFrequencies[0];
        if (n1 > 0 && castFrequencies.every(([, n2]) => n2 === n1)) {
            const items = castFrequencies.map(([t]) => t);
            dupChecker(items, (item) => item);
            return (rng) => () => items[Math.floor(items.length * rng())];
        }
    }
    else {
        const items = [...frequencies];
        if (items.length === 1) {
            const t = items[0];
            return (_rng) => () => t;
        }
        dupChecker(items, (item) => item);
        return (rng) => () => items[Math.floor(items.length * rng())];
    }
}
// remove impossible frequencies
function sanitize(frequencies) {
    return frequencies.map(([t, n]) => [t, Number(n)]).filter(([, n]) => Number.isFinite(n) && n > 0);
}
// make a function that converts a random number into an index, preserving the frequencies
// assuming the random number is from an even distribution in the interval [0, 1]
function thresholder(thresholds) {
    return eval(`(n) => ${portion(thresholds, 0, thresholds.length)}`);
}
// recursively convert the threshold list into nested ternary operators
// this allows us to find our index in O(log(n)) comparisons
function portion(thresholds, start, end) {
    const n = end - start;
    if (n === 1)
        return start.toString();
    const i = start + (n % 2 === 0 ? n / 2 - 1 : (n - 1) / 2);
    return `(n > ${thresholds[i]} ? ${portion(thresholds, i + 1, end)} : ${portion(thresholds, start, i + 1)})`;
}
/**
 * Generates a random number sequence generator. This is a demonstration generator
 * useful for testing and documentation. Any zero-argument function that returns
 * numbers will suffice for `pickMe`.
 *
 * The implementation of `rando` is taken from https://stackoverflow.com/a/47593316/15060051.
 * It is the Mulberry32 algorithm.
 *
 * Note, `rando` can be re-seeded by passing an optional number argument as the fresh seed (see
 * example below). This can be useful if you are sharing a random number generator among
 * several objects and you need to change random number sequences.
 *
 * @example
 * ```ts
 * import { rando } from 'pick-me'
 *
 * // make a random number generator with a seed of 1
 * const rng = rando(1)
 * const ar: number[] = []
 * for (let i = 0; i < 3; i++) ar.push(rng())
 * // => [ 0.6270739405881613, 0.002735721180215478, 0.5274470399599522 ]
 *
 * // the same seed will produce the same random number sequence
 * rando(1)()
 * // => 0.6270739405881613
 * rando(1)()
 * // => 0.6270739405881613
 *
 * // but you can provide a fresh seed to change the sequence
 * (rando(1) as any)(2)
 * // => 0.7342509443406016
 * ```
 *
 * @param {number} seed - the seed for a random number sequence
 * @returns {Rng} a generator of random numbers
 */
function rando(seed) {
    return (reseed) => {
        if (reseed !== undefined)
            seed = reseed;
        let t = (seed += 0x6d2b79f5);
        t = Math.imul(t ^ (t >>> 15), t | 1);
        t ^= t + Math.imul(t ^ (t >>> 7), t | 61);
        return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
    };
}
exports.rando = rando;


src/index.ts
/**
 * Generates a zero-argument function that returns an item from the frequency list
 * parameter with a probability concordant with its specified frequency.
 *
 * The required parameter is a list of pairs: an item to pick and its frequency. The "frequency"
 * is simply a positive number which determines the probability of picking one item relative to another.
 * If the parameters are `[[1, 1], [2, 2]]`, you will get twice as many 2s as 1s. You will get the same
 * results with `[[1, 0.1], [2, 0.2]]`, `[[1, 10], [2, 20]]`, and `[[1, 0.13], [2, 0.26]]`.
 *
 * The second, optional parameter is a random number sequence generator. Provide this if you want a reproducible
 * sequence of selections from your frequency list. If this doesn't concern you, leave it off and take
 * whatever `Math.random()` gives you.
 *
 * If the frequency list is not a list of pairs but just a list of items, this is equivalent to a list of
 * pairs where all the frequencies are the same. (Note, this will not work if the items are all pairs whose
 * second value is a number. If you want to pick among such pairs you'll need to wrap them in new pairs.)
 *
 * An item with
 * a non-positive frequency will be culled from the list. `pickMe` will throw an error if the list is empty.
 * Likewise, it will throw an error if any item appears more than once in the list. `[]` is
 * not good. `[[1, 1], [1, 2]]` is no better.
 *
 * Note, the item returned by `pickMe` is the item you provided, not a copy. If you are picking from a list of
 * objects and you modify one of them, it will remain modified.
 *
 * Finally, `pickMe` compiles an efficient picker from the parameters provided. As with other compiled things, such
 * as regular expressions, you want to do this one time in an initialization phase and then reuse it in loops.
 * The compilation is not terribly expensive, but skipping it altogether is cheaper.
 *
 * @example
 * ```ts
 * import { rando, pickMe } from 'pick-me'
 *
 * // make a random number generator with a seed of 1
 * const rng = rando(1)
 * const fooBarBaz = pickMe([["foo", 1], ["bar", 2], ["baz", 3]], rng)
 * const ar: string[] = []
 * for (let i = 0; i < 10; i++) ar.push(fooBarBaz())
 * // => ['bar', 'baz', 'bar', 'foo', 'foo', 'baz', 'bar', 'bar', 'baz', 'foo']
 * ```
 *
 * @param {[T, number][] | T[]} frequencies - phrases to match
 * @param {Rng} [rando=() => Math.random()] - random number generator
 * @returns {() => T} a picker of T's
 */
export function pickMe<T>(frequencies: [T, number][] | T[], rando?: Rng): () => T {
  const rng = rando ?? (() => Math.random())
  return pickMeToo(frequencies)(rng)
}

/**
 * Generates pickers of things. This separates the compilation step from frequencies so you
 * can plug different random number generators in.
 *
 * @example
 * ```ts
 * import { pickMeToo } from 'pick-me'
 *
 * const fooBarBaz = pickMeToo([["foo", 1], ["bar", 1000], ["baz", 1]])
 *
 * // minimum probablility generator; this will always return the same thing, which happens to be "foo"
 * fooBarBaz(() => 0)()
 * // => "bar"
 *
 * // maximum probability generator; this will always return the same thing, which happens to be  "baz"
 * fooBarBaz(() => 1)()
 * // => "baz"
 * ```
 *
 * @template T - the sort of thing to pick
 * @param {[T, number][] | T[]} frequencies - things to pick according to their frequencies
 * @returns {(rng: Rng) => () => T} - a function from random number generators to pickers of things in the frequency list
 */
export function pickMeToo<T>(frequencies: [T, number][] | T[]): (rng: Rng) => () => T {
  const simpleton = simplePicker(frequencies)
  if (simpleton) return simpleton

  frequencies = sanitize(frequencies as any as [T, number][])

  // short-circuit edge cases
  if (frequencies.length === 0) throw new Error(`nothing to select in ${frequencies}`)
  if (frequencies.length === 1) {
    const [t] = frequencies[0]
    return (_rng: Rng) => () => t
  }
  dupChecker(frequencies, (item: [T, number] | T) => (item as [T, number])[0])
  const [objects, nodes] = prepare(frequencies)
  const root = buildBranch(nodes)
  addThresholds(root, 0)
  const exp = ternarize(root)
  const picker: (n: number) => number = eval(`(p) => ${exp}`)
  return (rng: Rng) => () => objects[picker(rng())]
}

// convert a Node search tree into a nested ternary expression
function ternarize(n: Node): string {
  if (n.left && n.right) {
    return `(p < ${n.leftThreshold} ? ${ternarize(n.left)} : (p < ${n.rightThreshold} ? ${n.index} : ${ternarize(n.right)}))`
  } else if (n.left) {
    return `(p < ${n.leftThreshold} ? ${ternarize(n.left)} : ${n.index})`
  } else if (n.right) {
    return `(p < ${n.rightThreshold} ? ${n.index} : ${ternarize(n.right)})`
  } else {
    return n.index.toString()
  }
}

// convert the list of frequencies into Nodes (see below) and a list
// of things to pick
function prepare<T>(frequencies: [T, number][]): [T[], Node[]] {
  frequencies.sort(([, na], [,nb]) => nb - na)
  const sum = frequencies.reduce((acc, [, n]) => acc + n, 0)
  const objects: T[] = []
  const nodes: Node[] = []
  frequencies.forEach(([t, n], i) => {
    objects.push(t)
    nodes.push({interval: n / sum, index: i})
  })
  return [objects, nodes]
}

// a temporary data structure used for converting the list of frequencies
// into a tree structure we can search efficiently
type Node = {
  index: number,
  interval: number,
  sum?: number,
  leftThreshold?: number
  rightThreshold?: number
  left?: Node
  right?: Node
}

// given a search tree, add the probability thresholds for each node
function addThresholds(n: Node, acc: number) {
  if (n.left) {
    addThresholds(n.left, acc)
    acc += n.left.sum!
    n.leftThreshold = acc
  }
  if (n.right) {
    acc += n.interval
    n.rightThreshold = acc
    addThresholds(n.right, acc)
  }
}

// convert a list of sorted nodes into an efficiently searchable tree
function buildBranch(nodes: Node[]): Node {
  const sum = nodes.reduce((acc, n) => acc + n.interval, 0)
  const n = nodes.shift()!
  n.sum = sum
  if (nodes.length === 0) return n
  if (n.interval * 3 >= sum) {
    // the frequencies are so skewed a linear search is more efficient than a binary search
    n.right = buildBranch(nodes)
  } else {
    // perform a binary search, considering wide intervals before narrow ones
    const [left, right] = nodes.reduce(([left, right, idx], n) => {
      if (belongsToLeftBranch(idx)) left.push(n)
      else right.push(n)
      return [left, right, idx + 1]
    }, [[], [], 1] as [Node[], Node[], number])
    n.left = buildBranch(left)
    if (right.length > 0) n.right = buildBranch(right)
  }
  return n
}

// treating our sorted list of intervals as a heap, use the heap parent identification rules
// to determine whether the index in question is under the left or the right branch under the
// root node
function belongsToLeftBranch(i: number): boolean {
  if (i === 0) return true
  if (i < 0) return false
  return belongsToLeftBranch(Math.floor((i - 1) / 2))
}

/**
 * The type of random number generators. Ideally this will return number from a continuous distribution
 * in the interval [0, 1], but for `pickMe` and number less than 0 is the same as 0 and any number
 * greater than 1 is the same as 1.
 */
export type Rng = () => number

// throw an error in case of duplicates
// it is difficult to do this efficiently because we allow T to be anything, so we can't for instance use a Set
function dupChecker<T>(frequencies: [T, number][] | T[], toT: (item: [T, number] | T) => T) {
  for (let i = 0, lim = frequencies.length - 1; i < lim; i++) {
    const t = toT(frequencies[i])
    for (let j = i + 1; j < frequencies.length; j++) {
      if (t === toT(frequencies[j])) throw new Error(`${t} appears more than once among frequencies`)
    }
  }
}

// if all frequencies are the same, we can make a simpler picker
function simplePicker<T>(frequencies: [T, number][] | T[]): undefined | ((rng: Rng) => () => T) {
  if (frequencies.length === 0) return
  if ((frequencies as any).every((s: any) => Array.isArray(s) && s.length === 2 && typeof s[1] === 'number')) {
    const castFrequencies: [T, number][] = frequencies as any as [T, number][]
    const [, n1] = castFrequencies[0]
    if (n1 > 0 && castFrequencies.every(([, n2]) => n2 === n1)) {
      const items = castFrequencies.map(([t]) => t)
      dupChecker(items, (item: [T, number] | T) => item as T)
      return (rng: Rng) => () => items[Math.floor(items.length * rng())]
    }
  } else {
    const items: T[] = [...frequencies] as any as T[]
    if (items.length === 1) {
      const t = items[0]
      return (_rng: Rng) => () => t
    }
    dupChecker(items, (item: [T, number] | T) => item as T)
    return (rng: Rng) => () => items[Math.floor(items.length * rng())]
  }
}

// remove impossible frequencies
function sanitize<T>(frequencies: [T, number][]): [T, number][] {
  return frequencies.map(([t, n]) => [t, Number(n)]).filter(([, n]) => Number.isFinite(n) && n > 0) as any as [
    T,
    number,
  ][]
}

/**
 * Generates a random number sequence generator. This is a demonstration generator
 * useful for testing and documentation. Any zero-argument function that returns
 * numbers will suffice for `pickMe`.
 *
 * The implementation of `rando` is taken from https://stackoverflow.com/a/47593316/15060051.
 * It is the Mulberry32 algorithm.
 * 
 * Note, `rando` can be re-seeded by passing an optional number argument as the fresh seed (see
 * example below). This can be useful if you are sharing a random number generator among
 * several objects and you need to change random number sequences.
 *
 * @example
 * ```ts
 * import { rando } from 'pick-me'
 *
 * // make a random number generator with a seed of 1
 * const rng = rando(1)
 * const ar: number[] = []
 * for (let i = 0; i < 3; i++) ar.push(rng())
 * // => [ 0.6270739405881613, 0.002735721180215478, 0.5274470399599522 ]
 * 
 * // the same seed will produce the same random number sequence
 * rando(1)()
 * // => 0.6270739405881613
 * rando(1)()
 * // => 0.6270739405881613
 * 
 * // but you can provide a fresh seed to change the sequence
 * (rando(1) as any)(2)
 * // => 0.7342509443406016
 * ```
 *
 * @param {number} seed - the seed for a random number sequence
 * @returns {Rng} a generator of random numbers
 */
export function rando(seed: number): Rng {
  return (reseed?: number) => {
    if (reseed !== undefined) seed = reseed
    let t = (seed += 0x6d2b79f5)
    t = Math.imul(t ^ (t >>> 15), t | 1)
    t ^= t + Math.imul(t ^ (t >>> 7), t | 61)
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296
  }
}
